File: frontend/ios/App/App/NativeResponseCachePlugin.swift
import Capacitor
import CryptoKit
import Foundation

@objc(NativeResponseCachePlugin)
public class NativeResponseCachePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NativeResponseCachePlugin"
    public let jsName = "NativeResponseCache"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setEntry", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getEntry", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "invalidate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getStats", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "trim", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearAll", returnType: CAPPluginReturnPromise)
    ]

    private let queue = DispatchQueue(label: "com.xinya.native-response-cache", qos: .utility)
    private let cacheDirName = "native-response-cache"
    private let maxCacheBytes: Int64 = 20 * 1024 * 1024
    private var cacheRoot: URL?

    @objc override public func load() {
        try? ensureCacheRoot()
    }

    @objc func setEntry(_ call: CAPPluginCall) {
        guard let cacheKey = call.getString("cacheKey")?.trimmingCharacters(in: .whitespacesAndNewlines), !cacheKey.isEmpty else {
            call.reject("cacheKey is required")
            return
        }

        let url = call.getString("url", "")
        let status = call.getInt("status", 200)
        let statusText = call.getString("statusText", "OK")
        let headers = call.getObject("headers") ?? [:]
        let body = call.getString("body", "")

        queue.async { [weak self] in
            guard let self else { return }
            do {
                try self.writeEntry(
                    cacheKey: cacheKey,
                    url: url,
                    status: status,
                    statusText: statusText,
                    headers: headers,
                    body: body
                )
                self.resolve(call)
            } catch {
                self.reject(call, error, fallback: "setEntry failed")
            }
        }
    }

    @objc func getEntry(_ call: CAPPluginCall) {
        guard let cacheKey = call.getString("cacheKey")?.trimmingCharacters(in: .whitespacesAndNewlines), !cacheKey.isEmpty else {
            call.reject("cacheKey is required")
            return
        }

        queue.async { [weak self] in
            guard let self else { return }
            do {
                guard let entry = try self.readEntry(cacheKey) else {
                    self.resolve(call, ["exists": false])
                    return
                }
                self.resolve(call, [
                    "exists": true,
                    "url": entry.url,
                    "status": entry.status,
                    "statusText": entry.statusText,
                    "headers": entry.headers,
                    "body": entry.body,
                    "updatedAt": entry.updatedAt
                ])
            } catch {
                self.reject(call, error, fallback: "getEntry failed")
            }
        }
    }

    @objc func invalidate(_ call: CAPPluginCall) {
        let cacheKey = call.getString("cacheKey")?.trimmingCharacters(in: .whitespacesAndNewlines)
        let prefix = call.getString("prefix")?.trimmingCharacters(in: .whitespacesAndNewlines)
        if (cacheKey?.isEmpty ?? true) && (prefix?.isEmpty ?? true) {
            call.reject("cacheKey or prefix is required")
            return
        }

        queue.async { [weak self] in
            guard let self else { return }
            do {
                if let cacheKey, !cacheKey.isEmpty {
                    try self.deleteEntry(cacheKey)
                }
                if let prefix, !prefix.isEmpty {
                    try self.deleteEntriesByPrefix(prefix)
                }
                self.resolve(call)
            } catch {
                self.reject(call, error, fallback: "invalidate failed")
            }
        }
    }

    @objc func clearAll(_ call: CAPPluginCall) {
        queue.async { [weak self] in
            guard let self else { return }
            do {
                let root = try self.ensureCacheRoot()
                try? FileManager.default.removeItem(at: root)
                try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
                self.resolve(call)
            } catch {
                self.reject(call, error, fallback: "clearAll failed")
            }
        }
    }

    @objc func getStats(_ call: CAPPluginCall) {
        queue.async { [weak self] in
            guard let self else { return }
            do {
                let stats = try self.buildStats(maxBytes: self.maxCacheBytes, trimmedEntries: 0, trimmedBytes: 0)
                self.resolve(call, stats.result)
            } catch {
                self.reject(call, error, fallback: "getStats failed")
            }
        }
    }

    @objc func trim(_ call: CAPPluginCall) {
        let requestedMax = call.getDouble("maxBytes").map { Int64($0.rounded()) }
        let maxBytes = requestedMax.map { Swift.max(0, $0) } ?? maxCacheBytes

        queue.async { [weak self] in
            guard let self else { return }
            do {
                let stats = try self.trimToLimit(maxBytes)
                self.resolve(call, stats.result)
            } catch {
                self.reject(call, error, fallback: "trim failed")
            }
        }
    }

    private func writeEntry(
        cacheKey: String,
        url: String,
        status: Int,
        statusText: String,
        headers: JSObject,
        body: String
    ) throws {
        let bodyUrl = try bodyFile(cacheKey)
        let metaUrl = try metaFile(cacheKey)
        try Data(body.utf8).write(to: bodyUrl, options: .atomic)
        try writeMeta([
            "cacheKey": cacheKey,
            "url": url,
            "status": status,
            "statusText": statusText,
            "fileName": bodyUrl.lastPathComponent,
            "headers": headers,
            "updatedAt": responseCacheNowMilliseconds()
        ], to: metaUrl)
        try trimToLimit(maxCacheBytes)
    }

    private func readEntry(_ cacheKey: String) throws -> ResponseCacheEntry? {
        let metaUrl = try metaFile(cacheKey)
        guard var meta = readMeta(metaUrl) else {
            return nil
        }
        guard let bodyUrl = resolveBodyFile(meta), FileManager.default.fileExists(atPath: bodyUrl.path) else {
            try? FileManager.default.removeItem(at: metaUrl)
            return nil
        }
        let body = (try? String(contentsOf: bodyUrl, encoding: .utf8)) ?? ""
        let updatedAt = responseCacheNowMilliseconds()
        meta["updatedAt"] = updatedAt
        try? writeMeta(meta, to: metaUrl)

        return ResponseCacheEntry(
            url: meta["url"] as? String ?? "",
            status: responseCacheInt(meta["status"]) ?? 200,
            statusText: meta["statusText"] as? String ?? "OK",
            headers: normalizeHeaders(meta["headers"] as? [String: Any]),
            body: body,
            updatedAt: updatedAt
        )
    }

    private func deleteEntry(_ cacheKey: String) throws {
        let metaUrl = try metaFile(cacheKey)
        if let meta = readMeta(metaUrl), let bodyUrl = resolveBodyFile(meta) {
            try? FileManager.default.removeItem(at: bodyUrl)
        }
        try? FileManager.default.removeItem(at: metaUrl)
    }

    private func deleteEntriesByPrefix(_ prefix: String) throws {
        let root = try ensureCacheRoot()
        let files = try FileManager.default.contentsOfDirectory(at: root, includingPropertiesForKeys: nil)
        for file in files where file.pathExtension == "json" {
            guard let meta = readMeta(file) else {
                try? FileManager.default.removeItem(at: file)
                continue
            }
            if (meta["cacheKey"] as? String)?.hasPrefix(prefix) == true {
                if let bodyUrl = resolveBodyFile(meta) {
                    try? FileManager.default.removeItem(at: bodyUrl)
                }
                try? FileManager.default.removeItem(at: file)
            }
        }
    }

    private func metaFile(_ cacheKey: String) throws -> URL {
        try ensureCacheRoot().appendingPathComponent("\(responseCacheSha256(cacheKey)).json")
    }

    private func bodyFile(_ cacheKey: String) throws -> URL {
        try ensureCacheRoot().appendingPathComponent("\(responseCacheSha256(cacheKey)).body")
    }

    private func resolveBodyFile(_ meta: [String: Any]) -> URL? {
        guard let fileName = meta["fileName"] as? String, !fileName.isEmpty, let cacheRoot else {
            return nil
        }
        return cacheRoot.appendingPathComponent(fileName)
    }

    private func trimToLimit(_ maxBytes: Int64) throws -> ResponseCacheStats {
        var entries = try readCacheIndex()
        var totalBytes = entries.reduce(Int64(0)) { $0 + $1.size }

        if totalBytes <= maxBytes {
            return ResponseCacheStats(
                entryCount: entries.count,
                totalBytes: totalBytes,
                maxBytes: maxBytes,
                trimmedEntries: 0,
                trimmedBytes: 0
            )
        }

        entries.sort { $0.updatedAt < $1.updatedAt }
        var trimmedEntries = 0
        var trimmedBytes: Int64 = 0

        for entry in entries {
            if totalBytes <= maxBytes || entries.count - trimmedEntries <= 1 {
                break
            }

            try? FileManager.default.removeItem(at: entry.bodyUrl)
            try? FileManager.default.removeItem(at: entry.metaUrl)
            totalBytes -= entry.size
            trimmedBytes += entry.size
            trimmedEntries += 1
        }

        return try buildStats(maxBytes: maxBytes, trimmedEntries: trimmedEntries, trimmedBytes: trimmedBytes)
    }

    private func buildStats(maxBytes: Int64, trimmedEntries: Int, trimmedBytes: Int64) throws -> ResponseCacheStats {
        let entries = try readCacheIndex()
        let totalBytes = entries.reduce(Int64(0)) { $0 + $1.size }
        return ResponseCacheStats(
            entryCount: entries.count,
            totalBytes: totalBytes,
            maxBytes: maxBytes,
            trimmedEntries: trimmedEntries,
            trimmedBytes: trimmedBytes
        )
    }

    private func readCacheIndex() throws -> [ResponseCacheIndexEntry] {
        let root = try ensureCacheRoot()
        let files = try FileManager.default.contentsOfDirectory(at: root, includingPropertiesForKeys: nil)
        var entries: [ResponseCacheIndexEntry] = []

        for file in files where file.pathExtension == "json" {
            guard let meta = readMeta(file) else {
                try? FileManager.default.removeItem(at: file)
                continue
            }
            guard let bodyUrl = resolveBodyFile(meta), FileManager.default.fileExists(atPath: bodyUrl.path) else {
                try? FileManager.default.removeItem(at: file)
                continue
            }

            entries.append(ResponseCacheIndexEntry(
                metaUrl: file,
                bodyUrl: bodyUrl,
                size: Int64(fileSize(bodyUrl) + fileSize(file)),
                updatedAt: responseCacheInt64(meta["updatedAt"]) ?? 0
            ))
        }

        return entries
    }

    private func ensureCacheRoot() throws -> URL {
        if let cacheRoot {
            return cacheRoot
        }
        let root = try FileManager.default.url(
            for: .cachesDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        ).appendingPathComponent(cacheDirName, isDirectory: true)
        try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
        cacheRoot = root
        return root
    }

    private func readMeta(_ url: URL) -> [String: Any]? {
        guard let data = try? Data(contentsOf: url) else {
            return nil
        }
        return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
    }

    private func writeMeta(_ meta: [String: Any], to url: URL) throws {
        let data = try JSONSerialization.data(withJSONObject: meta)
        try data.write(to: url, options: .atomic)
    }

    private func fileSize(_ url: URL) -> Int {
        ((try? FileManager.default.attributesOfItem(atPath: url.path)[.size]) as? NSNumber)?.intValue ?? 0
    }

    private func normalizeHeaders(_ source: [String: Any]?) -> JSObject {
        var headers: JSObject = [:]
        for (key, value) in source ?? [:] {
            if value is NSNull {
                headers[key] = ""
                continue
            }
            headers[key] = String(describing: value)
        }
        return headers
    }

    private func resolve(_ call: CAPPluginCall, _ data: PluginCallResultData = [:]) {
        DispatchQueue.main.async {
            call.resolve(data)
        }
    }

    private func reject(_ call: CAPPluginCall, _ error: Error, fallback: String) {
        DispatchQueue.main.async {
            call.reject(error.localizedDescription.isEmpty ? fallback : error.localizedDescription)
        }
    }
}

private struct ResponseCacheEntry {
    let url: String
    let status: Int
    let statusText: String
    let headers: JSObject
    let body: String
    let updatedAt: Int
}

private struct ResponseCacheIndexEntry {
    let metaUrl: URL
    let bodyUrl: URL
    let size: Int64
    let updatedAt: Int64
}

private struct ResponseCacheStats {
    let entryCount: Int
    let totalBytes: Int64
    let maxBytes: Int64
    let trimmedEntries: Int
    let trimmedBytes: Int64

    var result: PluginCallResultData {
        [
            "entryCount": entryCount,
            "totalBytes": totalBytes,
            "maxBytes": maxBytes,
            "trimmedEntries": trimmedEntries,
            "trimmedBytes": trimmedBytes
        ]
    }
}

private func responseCacheSha256(_ value: String) -> String {
    let digest = SHA256.hash(data: Data(value.utf8))
    return digest.map { String(format: "%02x", $0) }.joined()
}

private func responseCacheNowMilliseconds() -> Int {
    Int(Date().timeIntervalSince1970 * 1000)
}

private func responseCacheInt(_ value: Any?) -> Int? {
    if value is NSNull { return nil }
    if let value = value as? Int { return value }
    if let value = value as? NSNumber { return value.intValue }
    if let value = value as? String { return Int(value) }
    return nil
}

private func responseCacheInt64(_ value: Any?) -> Int64? {
    if value is NSNull { return nil }
    if let value = value as? Int64 { return value }
    if let value = value as? Int { return Int64(value) }
    if let value = value as? NSNumber { return value.int64Value }
    if let value = value as? String { return Int64(value) }
    return nil
}

File: frontend/android/app/src/main/java/com/xinya/app/NativeResponseCachePlugin.java
package com.xinya.app;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONObject;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@CapacitorPlugin(name = "NativeResponseCache")
public class NativeResponseCachePlugin extends Plugin {

    private static final String CACHE_DIR_NAME = "native-response-cache";
    private static final long MAX_CACHE_BYTES = 20L * 1024L * 1024L;

    private final ExecutorService executor = Executors.newFixedThreadPool(2);
    private File cacheRoot;

    @Override
    public void load() {
        cacheRoot = new File(getContext().getFilesDir(), CACHE_DIR_NAME);
        if (!cacheRoot.exists()) {
            //noinspection ResultOfMethodCallIgnored
            cacheRoot.mkdirs();
        }
    }

    @Override
    protected void handleOnDestroy() {
        executor.shutdownNow();
    }

    @PluginMethod
    public void setEntry(PluginCall call) {
        String cacheKey = call.getString("cacheKey");
        String url = call.getString("url", "");
        Integer statusValue = call.getInt("status");
        int status = statusValue != null ? statusValue : 200;
        String statusText = call.getString("statusText", "OK");
        JSObject headers = call.getObject("headers");
        String body = call.getString("body", "");

        if (cacheKey == null || cacheKey.trim().isEmpty()) {
            call.reject("cacheKey is required");
            return;
        }

        final JSObject finalHeaders = headers != null ? headers : new JSObject();
        executor.execute(() -> {
            try {
                writeEntry(cacheKey, url, status, statusText, finalHeaders, body);
                dispatchToMainThread(call::resolve);
            } catch (Exception error) {
                String message = error.getMessage() != null ? error.getMessage() : "setEntry failed";
                dispatchToMainThread(() -> call.reject(message));
            }
        });
    }

    @PluginMethod
    public void getEntry(PluginCall call) {
        String cacheKey = call.getString("cacheKey");
        if (cacheKey == null || cacheKey.trim().isEmpty()) {
            call.reject("cacheKey is required");
            return;
        }

        executor.execute(() -> {
            try {
                CachedEntry entry = readEntry(cacheKey);
                JSObject result = new JSObject();
                if (entry == null) {
                    result.put("exists", false);
                    dispatchToMainThread(() -> call.resolve(result));
                    return;
                }

                result.put("exists", true);
                result.put("url", entry.url);
                result.put("status", entry.status);
                result.put("statusText", entry.statusText);
                result.put("headers", entry.headers);
                result.put("body", entry.body);
                result.put("updatedAt", entry.updatedAt);
                dispatchToMainThread(() -> call.resolve(result));
            } catch (Exception error) {
                String message = error.getMessage() != null ? error.getMessage() : "getEntry failed";
                dispatchToMainThread(() -> call.reject(message));
            }
        });
    }

    @PluginMethod
    public void invalidate(PluginCall call) {
        String cacheKey = call.getString("cacheKey");
        String prefix = call.getString("prefix");

        if ((cacheKey == null || cacheKey.trim().isEmpty()) && (prefix == null || prefix.trim().isEmpty())) {
            call.reject("cacheKey or prefix is required");
            return;
        }

        executor.execute(() -> {
            try {
                if (cacheKey != null && !cacheKey.trim().isEmpty()) {
                    deleteEntry(cacheKey);
                }
                if (prefix != null && !prefix.trim().isEmpty()) {
                    deleteEntriesByPrefix(prefix);
                }
                dispatchToMainThread(call::resolve);
            } catch (Exception error) {
                String message = error.getMessage() != null ? error.getMessage() : "invalidate failed";
                dispatchToMainThread(() -> call.reject(message));
            }
        });
    }

    @PluginMethod
    public void getStats(PluginCall call) {
        executor.execute(() -> {
            try {
                CacheStats stats = buildStats(MAX_CACHE_BYTES, 0, 0L);
                dispatchToMainThread(() -> call.resolve(stats.toJSObject()));
            } catch (Exception error) {
                String message = error.getMessage() != null ? error.getMessage() : "getStats failed";
                dispatchToMainThread(() -> call.reject(message));
            }
        });
    }

    @PluginMethod
    public void trim(PluginCall call) {
        Double maxBytesValue = call.getDouble("maxBytes");
        long maxBytes = maxBytesValue != null ? Math.max(0L, Math.round(maxBytesValue)) : MAX_CACHE_BYTES;

        executor.execute(() -> {
            try {
                CacheStats stats = trimToLimit(maxBytes);
                dispatchToMainThread(() -> call.resolve(stats.toJSObject()));
            } catch (Exception error) {
                String message = error.getMessage() != null ? error.getMessage() : "trim failed";
                dispatchToMainThread(() -> call.reject(message));
            }
        });
    }

    @PluginMethod
    public void clearAll(PluginCall call) {
        executor.execute(() -> {
            try {
                deleteRecursively(cacheRoot);
                if (!cacheRoot.exists()) {
                    //noinspection ResultOfMethodCallIgnored
                    cacheRoot.mkdirs();
                }
                dispatchToMainThread(call::resolve);
            } catch (Exception error) {
                String message = error.getMessage() != null ? error.getMessage() : "clearAll failed";
                dispatchToMainThread(() -> call.reject(message));
            }
        });
    }

    private void writeEntry(
        String cacheKey,
        String url,
        int status,
        String statusText,
        JSObject headers,
        String body
    ) throws Exception {
        ensureCacheRoot();

        File bodyFile = bodyFile(cacheKey);
        File metaFile = metaFile(cacheKey);

        try (OutputStream output = new FileOutputStream(bodyFile, false)) {
            output.write((body != null ? body : "").getBytes(StandardCharsets.UTF_8));
            output.flush();
        }

        JSONObject meta = new JSONObject();
        meta.put("cacheKey", cacheKey);
        meta.put("url", url != null ? url : "");
        meta.put("status", status);
        meta.put("statusText", statusText != null ? statusText : "OK");
        meta.put("fileName", bodyFile.getName());
        meta.put("headers", headers != null ? new JSONObject(headers.toString()) : new JSONObject());
        meta.put("updatedAt", System.currentTimeMillis());
        writeMeta(metaFile, meta);
        trimToLimit(MAX_CACHE_BYTES);
    }

    private CachedEntry readEntry(String cacheKey) {
        File metaFile = metaFile(cacheKey);
        JSONObject meta = readMeta(metaFile);
        if (meta == null) {
            return null;
        }

        File bodyFile = resolveBodyFile(meta);
        if (bodyFile == null || !bodyFile.exists()) {
            deleteQuietly(metaFile(cacheKey));
            return null;
        }

        long updatedAt = System.currentTimeMillis();
        try {
            meta.put("updatedAt", updatedAt);
            writeMeta(metaFile, meta);
        } catch (Exception ignored) {
            updatedAt = meta.optLong("updatedAt", 0L);
        }

        return new CachedEntry(
            meta.optString("url", ""),
            meta.optInt("status", 200),
            meta.optString("statusText", "OK"),
            readHeaders(meta.optJSONObject("headers")),
            readTextFile(bodyFile),
            updatedAt
        );
    }

    private void deleteEntry(String cacheKey) {
        JSONObject meta = readMeta(metaFile(cacheKey));
        if (meta != null) {
            File bodyFile = resolveBodyFile(meta);
            if (bodyFile != null) {
                deleteQuietly(bodyFile);
            }
        }
        deleteQuietly(metaFile(cacheKey));
    }

    private void deleteEntriesByPrefix(String prefix) {
        ensureCacheRoot();
        File[] files = cacheRoot.listFiles((dir, name) -> name.endsWith(".json"));
        if (files == null) {
            return;
        }

        for (File metaFile : files) {
            JSONObject meta = readMeta(metaFile);
            if (meta == null) {
                deleteQuietly(metaFile);
                continue;
            }

            if (meta.optString("cacheKey", "").startsWith(prefix)) {
                File bodyFile = resolveBodyFile(meta);
                if (bodyFile != null) {
                    deleteQuietly(bodyFile);
                }
                deleteQuietly(metaFile);
            }
        }
    }

    private File metaFile(String cacheKey) {
        return new File(cacheRoot, sha256(cacheKey) + ".json");
    }

    private File bodyFile(String cacheKey) {
        return new File(cacheRoot, sha256(cacheKey) + ".body");
    }

    private File resolveBodyFile(JSONObject meta) {
        String fileName = meta.optString("fileName", "");
        if (fileName.isEmpty()) {
            return null;
        }
        return new File(cacheRoot, fileName);
    }

    private CacheStats trimToLimit(long maxBytes) {
        List<CacheIndexEntry> entries = readCacheIndex();
        long totalBytes = 0L;
        for (CacheIndexEntry entry : entries) {
            totalBytes += entry.size;
        }

        if (totalBytes <= maxBytes) {
            return new CacheStats(entries.size(), totalBytes, maxBytes, 0, 0L);
        }

        entries.sort(Comparator.comparingLong(entry -> entry.updatedAt));
        int trimmedEntries = 0;
        long trimmedBytes = 0L;

        for (CacheIndexEntry entry : entries) {
            if (totalBytes <= maxBytes || entries.size() - trimmedEntries <= 1) {
                break;
            }

            long entrySize = entry.size;
            deleteQuietly(entry.bodyFile);
            deleteQuietly(entry.metaFile);
            totalBytes -= entrySize;
            trimmedBytes += entrySize;
            trimmedEntries += 1;
        }

        return buildStats(maxBytes, trimmedEntries, trimmedBytes);
    }

    private CacheStats buildStats(long maxBytes, int trimmedEntries, long trimmedBytes) {
        List<CacheIndexEntry> entries = readCacheIndex();
        long totalBytes = 0L;
        for (CacheIndexEntry entry : entries) {
            totalBytes += entry.size;
        }
        return new CacheStats(entries.size(), totalBytes, maxBytes, trimmedEntries, trimmedBytes);
    }

    private List<CacheIndexEntry> readCacheIndex() {
        ensureCacheRoot();
        List<CacheIndexEntry> entries = new ArrayList<>();
        File[] files = cacheRoot.listFiles((dir, name) -> name.endsWith(".json"));
        if (files == null) {
            return entries;
        }

        for (File metaFile : files) {
            JSONObject meta = readMeta(metaFile);
            if (meta == null) {
                deleteQuietly(metaFile);
                continue;
            }

            File bodyFile = resolveBodyFile(meta);
            if (bodyFile == null || !bodyFile.exists()) {
                deleteQuietly(metaFile);
                continue;
            }

            long size = bodyFile.length() + metaFile.length();
            long updatedAt = meta.optLong("updatedAt", 0L);
            entries.add(new CacheIndexEntry(metaFile, bodyFile, size, updatedAt));
        }

        return entries;
    }

    private void ensureCacheRoot() {
        if (cacheRoot == null) {
            cacheRoot = new File(getContext().getFilesDir(), CACHE_DIR_NAME);
        }
        if (!cacheRoot.exists()) {
            //noinspection ResultOfMethodCallIgnored
            cacheRoot.mkdirs();
        }
    }

    private JSONObject readMeta(File file) {
        try {
            if (file == null || !file.exists()) {
                return null;
            }
            try (InputStream input = new FileInputStream(file)) {
                byte[] buffer = new byte[(int) file.length()];
                int read = input.read(buffer);
                if (read <= 0) {
                    return null;
                }
                return new JSONObject(new String(buffer, 0, read, StandardCharsets.UTF_8));
            }
        } catch (Exception ignored) {
            return null;
        }
    }

    private String readTextFile(File file) {
        try (InputStream input = new FileInputStream(file)) {
            byte[] buffer = new byte[(int) file.length()];
            int read = input.read(buffer);
            if (read <= 0) {
                return "";
            }
            return new String(buffer, 0, read, StandardCharsets.UTF_8);
        } catch (Exception ignored) {
            return "";
        }
    }

    private JSObject readHeaders(JSONObject source) {
        JSObject headers = new JSObject();
        if (source == null) {
            return headers;
        }

        Iterator<String> keys = source.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            headers.put(key, source.optString(key, ""));
        }
        return headers;
    }

    private void writeMeta(File metaFile, JSONObject meta) throws Exception {
        try (OutputStream output = new FileOutputStream(metaFile, false)) {
            output.write(meta.toString().getBytes(StandardCharsets.UTF_8));
            output.flush();
        }
    }

    private String sha256(String value) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hashed = digest.digest(value.getBytes(StandardCharsets.UTF_8));
            StringBuilder builder = new StringBuilder();
            for (byte current : hashed) {
                builder.append(String.format(Locale.US, "%02x", current));
            }
            return builder.toString();
        } catch (Exception error) {
            throw new IllegalStateException("Unable to hash cache key", error);
        }
    }

    private void deleteQuietly(File file) {
        if (file != null && file.exists()) {
            //noinspection ResultOfMethodCallIgnored
            file.delete();
        }
    }

    private void deleteRecursively(File file) {
        if (file == null || !file.exists()) {
            return;
        }

        if (file.isDirectory()) {
            File[] children = file.listFiles();
            if (children != null) {
                for (File child : children) {
                    deleteRecursively(child);
                }
            }
        }

        //noinspection ResultOfMethodCallIgnored
        file.delete();
    }

    private void dispatchToMainThread(Runnable action) {
        if (getActivity() != null) {
            getActivity().runOnUiThread(action);
            return;
        }
        action.run();
    }

    private static final class CacheStats {
        final int entryCount;
        final long totalBytes;
        final long maxBytes;
        final int trimmedEntries;
        final long trimmedBytes;

        CacheStats(int entryCount, long totalBytes, long maxBytes, int trimmedEntries, long trimmedBytes) {
            this.entryCount = entryCount;
            this.totalBytes = totalBytes;
            this.maxBytes = maxBytes;
            this.trimmedEntries = trimmedEntries;
            this.trimmedBytes = trimmedBytes;
        }

        JSObject toJSObject() {
            JSObject object = new JSObject();
            object.put("entryCount", entryCount);
            object.put("totalBytes", totalBytes);
            object.put("maxBytes", maxBytes);
            object.put("trimmedEntries", trimmedEntries);
            object.put("trimmedBytes", trimmedBytes);
            return object;
        }
    }

    private static final class CacheIndexEntry {
        final File metaFile;
        final File bodyFile;
        final long size;
        final long updatedAt;

        CacheIndexEntry(File metaFile, File bodyFile, long size, long updatedAt) {
            this.metaFile = metaFile;
            this.bodyFile = bodyFile;
            this.size = size;
            this.updatedAt = updatedAt;
        }
    }

    private static final class CachedEntry {
        final String url;
        final int status;
        final String statusText;
        final JSObject headers;
        final String body;
        final long updatedAt;

        CachedEntry(String url, int status, String statusText, JSObject headers, String body, long updatedAt) {
            this.url = url;
            this.status = status;
            this.statusText = statusText;
            this.headers = headers;
            this.body = body;
            this.updatedAt = updatedAt;
        }
    }
}
